src/parser/content-parser.php

```diff
@@ -189,37 +196,114 @@ public function parse( $post_content, $post_id = null, $filter_options = [] ) {
 		}
 	}
 
+	/**
+	 * Helper function to render a parsed block, so that we can benefit from
+	 * core-powered functions like block bindings and synced patterns.
+	 *
+	 * This loosely mirrors the code in the `render_block` function in core, but
+	 * allows us to capture the block instance so that we can traverse the tree:
+	 * https://github.com/WordPress/WordPress/blob/01d2199622d52b08d1704871770c68e35d2a80dc/wp-includes/blocks.php#L2012
+	 *
+	 * @param array $parsed_block Parsed block (result of `parse_blocks`).
+	 * @return WP_Block
+	 */
+	protected function render_parsed_block( array $parsed_block ): WP_Block {
+		$context = [];
+		if ( is_int( $this->post_id ) ) {
+			$context['postId']   = $this->post_id;
+			$context['postType'] = get_post_type( $this->post_id );
+		}
+
+		$context = apply_filters( 'render_block_context', $context, $parsed_block, null );
+
+		$block_instance = new WP_Block( $parsed_block, $context, $this->block_registry );
+		$block_instance->render();
+
+		return $block_instance;
+	}
+
 	/**
 	 * Processes a single block, and returns the sourced block data.
 	 *
-	 * @param array           $block Block to be processed.
-	 * @param WP_Block_Type[] $registered_blocks Blocks that have been registered.
-	 * @param array           $filter_options Options to filter using, if any.
+	 * @param WP_Block $block          Block to be processed.
+	 * @param array    $filter_options Options to filter using, if any.
 	 *
 	 * @return array|null
 	 *
 	 * @access private
 	 */
-	protected function source_block( $block, $registered_blocks, $filter_options ) {
-		$block_name = $block['blockName'];
+	protected function source_block( WP_Block $block, array $filter_options ) {
+		$block_name = $block->name;
 
-		if ( ! $this->should_block_be_included( $block, $block_name, $filter_options ) ) {
+		if ( ! $this->should_block_be_included( $block, $filter_options ) ) {
 			return null;
 		}
 
-		if ( ! isset( $registered_blocks[ $block_name ] ) ) {
+		if ( ! $this->block_registry->is_registered( $block_name ) ) {
 			$this->add_missing_block_warning( $block_name );
 		}
 
-		$block_definition            = $registered_blocks[ $block_name ] ?? null;
-		$block_definition_attributes = $block_definition->attributes ?? [];
+		$sourced_block = [
+			'name'       => $block->name,
+			'attributes' => $this->apply_sourced_attributes( $block ),
+		];
 
-		$block_attributes = $block['attrs'];
+		// WP_Block#inner_blocks can be an array or WP_Block_List (iterable).
+		$inner_blocks = iterator_to_array( $block->inner_blocks );
+
+		// Recursively iterate over inner blocks.
+		$sourced_inner_blocks = array_values( array_filter( array_map( function ( $inner_block ) use ( $filter_options ) {
+			return $this->source_block( $inner_block, $filter_options );
+		}, $inner_blocks ) ) );
+
+		// This empty check is not strictly necessary, but previous versions did
+		// not set innerBlocks if the array was empty.
```

Comment:
Nit, I'd remove the "This empty check is not strictly necessary" part:

```suggestion
		// Only set innerBlocks if entries are present to match prior version behavior
```

Reply:
I took your suggestion in https://github.com/Automattic/vip-block-data-api/pull/74/commits/674d006f58d1a1ae8cb870f31c46035ce26c5752. 

I tend to think conditional API properties are an anti-pattern. As an API consumer I'd much rather see a consistent response shape. Making this change would require updating many test fixtures, but I'd be happy to do so it is viewed as an improvement. I doubt it would be a breaking change for any users.

Reply:
> I tend to think conditional API properties are an anti-pattern. As an API consumer I'd much rather see a consistent response shape.

That's a good point I hadn't considered. For now, let's keep the response shape as-is in case folks have strict equality checks in consumers like `if ( block.innerBlocks === undefined )`. I'd like to keep this change in mind if/when we make a breaking change to a `v2` API.